7#2 ログインページのMVC、CSSの作成

terukyou opened the pull request:
* MembersController.phpの26行目付近のreturn $this->redirect(['controller' => 'error']);はログイン状態だとエラーに飛ぶ設定なのでコードレビューの際はコメント化してください。
* feature/hinappu-7/#55は、会員登録のマージ前に作成したものなので新たにfeature/hinappu-7/#55のコピーを作成しました

hirayama-sachi committed:
add: ログイン機能初回作成

kazuyanakata commented on html/mycakeapp/src/Controller/MembersController.php:
```diff
@@ -15,8 +19,12 @@ class MembersController extends AppController
     {
-        $member = null;
-        $this->set('member', $member);
+        parent::initialize();
+        $member = $this->Auth->user();
+        if (!empty($member)) {
+            // コードレビュー時はコメント化してください
+            return $this->redirect(['controller' => 'error']);
+        }
```
①ログイン後にログアウトボタンを押すとエラーコントローラーに遷移してしまいます。
```suggestion
        parent::initialize();
        $member = $this->Auth->user();
        if (!empty($member) && $this->request->action !== 'logout') {
            // コードレビュー時はコメント化してください
            return $this->redirect(['controller' => 'error']);
        }
```
②他のコントローラー(main, mypage, reserves)でも同様な形でログインの有無による画面遷移の対応をお願いします
③エラー画面の作成もお願いいたします。(簡単なビュー作成だけなので時間はかからないかと！)

terukyou replied:
確かにログアウトにいくとエラー画面に遷移しますね。
修正します！！
同時に他のコントローラーもユーザーの状態によって遷移するよう変更し、エラー画面の作成も行います

kazuyanakata commented on html/mycakeapp/src/Controller/MembersController.php:
```diff
@@ -160,4 +168,39 @@ public function changed()
+            // email形式チェックのエラーメッセージがない時のみ認証失敗メッセージ(email)を追加する
+            $emailErrors = $entity->getError("email");
+            if (empty($emailErrors["email"])) {
+                $emailError = 'メールアドレスが間違っているようです';
```
<提案>
エラー文を「メールアドレスかパスワードまたはその両方が間違っているようです」に変えるのはどうでしょうか？
もし片方はあってた場合、両方間違えてるみたいな言い方になっていて混乱すると思うので！
```suggestion
                $emailError = 'メールアドレスかパスワードまたはその両方が間違っているようです';
```

terukyou replied:
バリデーションではじかれたのか、認証ではじかれたのか明確に分かっていいと思います！！
こちらも、修正いたします🙇‍♂️🙆‍♂️

kazuyanakata commented on html/mycakeapp/src/Controller/MembersController.php:
```diff
@@ -160,4 +168,39 @@ public function changed()
+            } else {
+                $emailError = null;
+            }
+            $passwordError = 'パスワードが間違っているようです';
```
上に同じくです！

kazuyanakata commented on html/mycakeapp/src/Template/Members/login.ctp:
```diff
@@ -0,0 +1,32 @@
+?>
+<div class="not-able-login">
+    <?php
+    echo $this->Html->link('会員登録', ['action' => 'create']);
+    echo $this->Html->link('パスワードを忘れた方はコチラ', ['action' => 'password']);
+    ?>
```
<提案>
この部分UXのことを考えて
ホバーしたら背景変更もしくは下線表示させるのはどうでしょうか？

terukyou replied:
おっしゃる通りだと思います！！
修正します！！🙇‍♂️🙆‍♂️

kazuyanakata reviewed (commented):
お疲れ様です！
ブランチの対応ありがとうございました！

4件コメント入れたのでご確認ください！中には提案も入っているのでご意見いただければなと思います！
よろしくお願いいたします！

hirayama-sachi committed:
fix: ログイン状態でログアウトにいくときはエラー画面に遷移しないよう修正

全てのmembersControllerがログイン状態だとエラー画面に飛ぶようになっていたので、ログアウトのみエラーに行かないよう修正

hirayama-sachi committed:
add: ログイン判定によっての画面遷移機能追加(コントローラー[mypage,reserves])

ログイン状態でないとエラー画面へ飛ばす機能を追加(seatはログイン画面に遷移させるため必要があるため除外)

hirayama-sachi committed:
add: エラーページの作成

エラーページのMVCとCSSを作成

hirayama-sachi committed:
fix: 認証失敗のメッセージを統一するよう修正

emailのエラーメッセージがemailの認証失敗メッセージと同一であることによる混乱を避けるため、
認証失敗メッセージをメールアドレスとパスワードのどちらにも使用できるようなものに変更

hirayama-sachi committed:
add: リンクをホバーすると下線が表示される機能追加

UX向上のため、会員登録とパスワード再登録へのリンクにホバー時に下線を表示する機能を追加

hirayama-sachi committed:
fix: 不必要なコードを削除(MembersController下のbeforeFilter())

MembersControllerではbeforeFilter()をオーバーライドしないので削除

hirayama-sachi committed:
fix: エラーページに行く際認証が必要になるバグを修正

エラーコントローラでAppControllerのbeforeFilter()を呼び忘れていたので追記した

kazuyanakata reviewed (commented):
LGTMです！！！
